keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,314 @@
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to extend the APIs for the status of LocalQueue and ClusterQueue
+to expose the order to workloads. The order will be only exposed up to some
+configurable depth.
+
+The approach of keeping the information in the status of
+the queues will allow to expose the information without an extra cost of new
+API requests, as the statuses are already updated on changes for the number of
+pending workloads.
+
+In order to keep the size of the information constrained only the head of the
+queue of pending workloads will be exposed.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user of Kueue with LocalQueue visibility I would like to know the
+position of my workload in the local queue. Knowing the position would allow me
+to compute the estimate arrival time (ETA) of my workload.
+
+I would like to be able to get this information by inspecting the local queue
+status.
+
+<!--
+#### Story 2
+
+As a user of Kueue with ClusterQueue visibility I would like to know
+the position of my workload in the cluster queue, so that I can estimate the
+ETA of my workload.
+
+I would like to be able to get this information by inspecting the cluster queue
+status.
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Local Queue API
+
+```golang
+// PendingWorkload contains the information identifying a pending workload in
+// the local queue.
+type PendingWorkload struct {
+	Name string
+
+	// Namespace indicates the namespace of the pending workload
+	Namespace string
+
+	// Position indicates the position of the workload among all pending
+	// workloads in the cluster queue.
+	Position *int32
+}
+
+type PendingWorkloadsStatus struct {
```

Comment:
Is there a reason to have this one field struct? can't we just put `TopList []PendingWorkload`  in the status?

Reply:
Yeah, I'm hesitant about this. I was thinking it might be useful to add here some other global information regarding pending workloads. For example, their total number in the cluster queue. I think if this structure was first we could also put the `pendingWorkloads` as `count` under it, but it is already outside, so we may just continue putting things at top level.

Reply:
maybe we could add `lastTranstionTimestamp` to indicate the last time it was updated or positions changed.

Reply:
Added, named as `LastChangeTime`

Reply:
Resolving as with the new field added the structure is a natural container for both the list and the `LastChangeTime`.